File: cpp/src/search_algorithms/cabs.cpp
#include "jpm/search_algorithm.hpp"

#include "jpm/search_option_utils.hpp"
#include "jpm/search_support.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <memory>
#include <string>
#include <vector>

namespace {

SearchResult make_result(
    bool solvable,
    const std::string &running,
    std::vector<std::string> plan,
    int path_length,
    const SearchStats &stats,
    const std::string &search_name) {
    SearchResult result;
    result.solvable = solvable;
    result.running = running;
    result.plan = std::move(plan);
    result.path_length = path_length;
    result.expanded = stats.expanded;
    result.generated = stats.generated;
    result.pruned = stats.pruned();
    result.pruned_by_unknown = stats.pruned_by_unknown;
    result.pruned_by_visited = stats.pruned_by_visited;
    result.goal_checked = stats.goal_checked;
    result.search = search_name;
    return result;
}

class Cabs final : public SearchAlgorithm {
public:
    const std::string &name() const override {
        static const std::string value = "cabs";
        return value;
    }

    SearchResult search(Task &task, const SearchOptions &options) const override {
        const int initial_beam_size =
            read_int_option(options.search_options, "initial_beam_size", 1, 1);
        const int max_beam_size =
            read_int_option(options.search_options, "max_beam_size", 64, initial_beam_size);
        const bool keep_all_layers =
            read_bool_option(options.search_options, "keep_all_layers", false);

        SearchStats stats;
        stats.generated = 0;

        auto start = std::chrono::steady_clock::now();
        int beam_size = initial_beam_size;
        std::vector<std::string> best_actions;
        bool has_best_actions = false;

        while (beam_size <= max_beam_size) {
            const double elapsed_seconds =
                std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count();
            const int remaining_seconds =
                static_cast<int>(static_cast<double>(options.timeout_seconds) - elapsed_seconds);
            if (remaining_seconds <= 0) {
                if (has_best_actions) {
                    return make_result(
                        true,
                        "SUCC",
                        best_actions,
                        static_cast<int>(best_actions.size()),
                        stats,
                        name());
                }
                return make_result(false, "TIMEOUT", {}, 0, stats, name());
            }

            const int runs_remaining =
                static_cast<int>(std::log2(static_cast<double>(max_beam_size) / beam_size)) + 1;
            const int run_time_limit = std::max(1, remaining_seconds / runs_remaining);

            SearchOptions beam_options = options;
            beam_options.algorithm = "beamsearch";
            beam_options.timeout_seconds = run_time_limit;
            beam_options.search_options["beam_size"] = json::Value{static_cast<double>(beam_size)};
            beam_options.search_options["keep_all_layers"] = json::Value{keep_all_layers};

            std::unique_ptr<SearchAlgorithm> solver = make_search_algorithm("beamsearch");
            SearchResult solver_result = solver->search(task, beam_options);

            stats.expanded += solver_result.expanded;
            stats.goal_checked += solver_result.goal_checked;
            stats.generated += solver_result.generated;
            stats.pruned_by_unknown += solver_result.pruned_by_unknown;
            stats.pruned_by_visited += solver_result.pruned_by_visited;

            if (solver_result.running == "SUCC") {
                if (!has_best_actions || solver_result.plan.size() < best_actions.size()) {
                    best_actions = solver_result.plan;
                    has_best_actions = true;
                }
            } else if (solver_result.running == "TIMEOUT") {
                beam_size *= 2;
                continue;
            }

            beam_size *= 2;
        }

        if (has_best_actions) {
            return make_result(
                true,
                "SUCC",
                best_actions,
                static_cast<int>(best_actions.size()),
                stats,
                name());
        }

        return make_result(false, "UNSOLVED", {}, 0, stats, name());
    }
};

std::unique_ptr<SearchAlgorithm> make_cabs() {
    return std::make_unique<Cabs>();
}

const SearchAlgorithmRegistration registration("cabs", make_cabs);

}  // namespace

File: jp/search_algorithms/cabs.py
from __future__ import annotations

import datetime
import math

from search_algorithm_config import read_bool_option, read_int_option, normalise_search_options
from search_algorithms.beam_search import beam_search
from search_core import BestFirstSearchEngine


class cabs(BestFirstSearchEngine):
    def __init__(self, handlers, search_name, search_options=None):
        super().__init__(handlers, search_name)
        self.search_options = normalise_search_options(search_options)
        self.initial_beam_size = read_int_option(
            self.search_options,
            "initial_beam_size",
            1,
            minimum=1,
        )
        self.max_beam_size = read_int_option(
            self.search_options,
            "max_beam_size",
            64,
            minimum=self.initial_beam_size,
        )
        self.keep_all_layers = read_bool_option(
            self.search_options,
            "keep_all_layers",
            False,
        )

    def searching(self, problem, time_out: int, memory_out: int):
        self._reset_search_state()
        self._initialise_problem_context(problem, time_out, memory_out)
        self.generated = 0

        start_time = datetime.datetime.now()
        beam_size = self.initial_beam_size
        best_actions: list[str] | None = None
        beam_sizes: list[int] = []

        while beam_size <= self.max_beam_size:
            elapsed_seconds = (datetime.datetime.now() - start_time).total_seconds()
            remaining_seconds = int(self.timeout.total_seconds() - elapsed_seconds)
            if remaining_seconds <= 0:
                if best_actions is not None:
                    self.result.update({"beam_runs": len(beam_sizes)})
                    self.result.update({"beam_sizes": beam_sizes})
                    return self._complete_result_from_actions(problem, "SUCC", best_actions, True)
                self.logger.info(
                    "Problem cannot be solved in the given time (%s).",
                    self.timeout.seconds,
                )
                self.result.update({"beam_runs": len(beam_sizes)})
                self.result.update({"beam_sizes": beam_sizes})
                return self._complete_result_from_actions(problem, "TIMEOUT", [], False)

            runs_remaining = int(math.log2(self.max_beam_size / beam_size)) + 1
            run_time_limit = max(1, remaining_seconds // runs_remaining)

            solver = beam_search(
                self.logger.handlers,
                f"{self.search_name}_beam_{beam_size}",
                self.search_options,
            )
            solver.beam_size = beam_size
            solver.keep_all_layers = self.keep_all_layers
            solver_result = solver.searching(problem, run_time_limit, memory_out)

            beam_sizes.append(beam_size)
            self.expanded += solver.expanded
            self.goal_checked += solver.goal_checked
            self.generated += solver.generated
            self.pruned_by_unknown += solver.pruned_by_unknown
            self.pruned_by_visited += solver.pruned_by_visited
            self.branch_factors.extend(solver.branch_factors)

            if solver_result.get("running") == "SUCC":
                actions = solver_result.get("plan", [])
                if best_actions is None or len(actions) < len(best_actions):
                    best_actions = actions
            elif solver_result.get("running") == "TIMEOUT":
                beam_size *= 2
                continue
            elif solver_result.get("running") == "MEMORYOUT":
                if best_actions is not None:
                    self.result.update({"beam_runs": len(beam_sizes)})
                    self.result.update({"beam_sizes": beam_sizes})
                    return self._complete_result_from_actions(problem, "SUCC", best_actions, True)
                self.logger.info(
                    "Beam search wrapper stopped with status %s.",
                    solver_result.get("running"),
                )
                self.result.update({"beam_runs": len(beam_sizes)})
                self.result.update({"beam_sizes": beam_sizes})
                return self._complete_result_from_actions(
                    problem,
                    solver_result.get("running"),
                    [],
                    False,
                )

            beam_size *= 2

        self.result.update({"beam_runs": len(beam_sizes)})
        self.result.update({"beam_sizes": beam_sizes})
        if best_actions is not None:
            return self._complete_result_from_actions(problem, "SUCC", best_actions, True)

        self.logger.info("Problem is not solvable")
        return self._complete_result_from_actions(problem, "UNSOLVED", [], False)